### PR TITLE
Make autoencoder API parameters match and add some input validation

### PR DESCRIPTION
**Pull request by ryanjdillon**
In this PR the following was done:
* enc_* and dec_* parameters in the feedforward factory were renamed to match the lstm factory arguments encoding_*, decoding_*
* Extra validation of input was added in factory utils to give a warning if only one of these layer configuraton parameters is passed.
* Defaults are now set after initial checking the passing of a single layer configuration parameter.
* Other validation statements in factories are now moved to utils.
* Tests were added to check these

### CODE REVIEW

File: gordo_components/model/factories/feedforward_autoencoder.py [hunk 1]
```diff
@@ -1,24 +1,27 @@
...
+from typing import Tuple, Dict, Any
 
 from keras import regularizers
 from keras.layers import Dense
 import keras
 
 from keras.models import Sequential as KerasSequential
 from gordo_components.model.register import register_model_builder
-from gordo_components.model.factories.model_factories_utils import hourglass_calc_dims
+from gordo_components.model.factories.model_factories_utils import (
```
Comment (milesgranger) [inline gordo_components/model/factories/feedforward_autoencoder.py]: To follow the same pattern, you can just name this `utils`; as that's what we have in most of our component modules.

File: gordo_components/model/factories/lstm_autoencoder.py [hunk 2]
```diff
@@ -71,17 +74,13 @@ class (e.x. Adam(lr=0.01,beta_1=0.9, beta_2=0.999)).  If no arguments are
     """
     n_features_out = n_features_out or n_features
 
-    if len(encoding_dim) != len(encoding_func):
-        raise ValueError(
-            f"Number of layers ({len(encoding_dim)}) and number of functions ({len(encoding_func)}) "
-            f"must be equal for the encoder."
-        )
+    encoding_dim, encoding_func = set_dim_func(
+        encoding_dim, encoding_func, (256, 128, 64), ("tanh", "tanh", "tanh")
```
Comment (milesgranger) [inline gordo_components/model/factories/lstm_autoencoder.py]: Rather than having the defaults down here, why not have them set in the function signature?

File: gordo_components/model/factories/model_factories_utils.py [hunk 3]
```diff
@@ -15,3 +16,36 @@ def hourglass_calc_dims(compression_factor, encoding_layers, n_features):
         round(n_features - (i * average_slope)) for i in range(1, encoding_layers + 1)
     ]
     return dims
+
+
+def _check_if_only_dim_or_func(dim, func):
```
Comment (milesgranger) [inline gordo_components/model/factories/model_factories_utils.py]: This PR isn't marked with `WIP` so I'll go ahead and say that for all these functions, we'll need type hints and numpy style docs for them.

File: tests/gordo_components/model/test_model_factories_utils.py [hunk 4]
```diff
@@ -1,25 +1,55 @@
...
+
+def test_check_dim_func_len():
+    # Raise a ValueError if the len of dim and func are not equal
+    with pytest.raises(ValueError):
+        _check_dim_func_len(dim=(256,128), func=("tanh", "tanh", "tanh"))
+
+
+@pytest.mark.parametrize("dim", ((256, 128, 64), (), (256, 128, 64)))
+@pytest.mark.parametrize("func", (("tanh", "tanh", "tanh"), ("tanh", "tanh", "tanh"), ()))
+def test_set_dim_func(dim, func):
```
Comment (milesgranger) [inline tests/gordo_components/model/test_model_factories_utils.py]: Short human readable description of the test goal, pretty please.

File: tests/gordo_components/model/test_model_factories_utils.py [hunk 5]
```diff
@@ -1,25 +1,55 @@
...
+    dims = hourglass_calc_dims(0.5, 3, 10)
+    assert dims == [8, 7, 5]
+    dims = hourglass_calc_dims(0.5, 3, 3)
+    assert dims == [3, 2, 2]
+    dims = hourglass_calc_dims(0.3, 3, 10)
+    assert dims == [8, 5, 3]
+    dims = hourglass_calc_dims(1, 3, 10)
+    assert dims == [10, 10, 10]
+    dims = hourglass_calc_dims(0, 3, 100000)
+    assert dims == [66667, 33334, 1]
```
Comment (milesgranger) [inline tests/gordo_components/model/test_model_factories_utils.py]: Nice, but I think this is a prime candidate for `input, expected` parameterization.

**Review by milesgranger — COMMENTED** [review]
Looking pretty good.

You'll need to reword all your commits to be in the imperative style we use. [good reference here](https://chris.beams.io/posts/git-commit/)

File: gordo_components/model/factories/lstm_autoencoder.py (hunk 2 above)
Comment (ryanjdillon) [reply gordo_components/model/factories/lstm_autoencoder.py]: Yeah, this didn't feel great. This was done to address the comment mentioned [here](https://github.com/equinor/gordo-components/issues/130#issuecomment-467005835). TL;DR Passing only dim or func with a different number of layers from the default will break. 

My solution here was to first check this and throw a warning if only one is passed, set the default to the one that wasn't passed, and then allow it to break as before if they (the one passed and the default used for the other) are of unequal length.

File: tests/gordo_components/model/test_model_factories_utils.py (hunk 5 above)
Comment (ryanjdillon) [reply tests/gordo_components/model/test_model_factories_utils.py]: This was the one test method for the pre-existing utils module that I put my routines in. Should I rewrite things like this as I come across them and include them in my PR, or maybe create a new issue/PR with things I uncover as I work on something else?

File: gordo_components/model/factories/feedforward_autoencoder.py (hunk 1 above)
Comment (ryanjdillon) [reply gordo_components/model/factories/feedforward_autoencoder.py]: This goes along with the last comment, but this was already there. Is it best for me to rename it as I come across it and include it in my PR, create a new issue/PR, etc?

File: gordo_components/model/factories/lstm_autoencoder.py (hunk 2 above)
Comment (ryanjdillon) [reply gordo_components/model/factories/lstm_autoencoder.py]: Going to roll this back and keep them in the signature with a less descriptive error if one not passed. :+1:

File: tests/gordo_components/model/test_model_factories_utils.py (hunk 5 above)
Comment (ryanjdillon) [reply tests/gordo_components/model/test_model_factories_utils.py]: I've included this in this PR, but just let me know if I should pull it out to a different PR.

Comment (milesgranger) [reply tests/gordo_components/model/test_model_factories_utils.py]: ya, just because it's a pre-existing condition doesn't mean we should repeat it. ;-)

File: gordo_components/model/factories/lstm_autoencoder.py [hunk 6]
```diff
@@ -41,16 +44,16 @@ def lstm_model(
...
+    encoding_dim: tuple
+        Tuple of numbers with the number of neurons in the encoding part.
+    decoding_dim: tuple
+        Tuple of numbers with the number of neurons in the decoding part.
+    encoding_func: tuple
         Activation functions for the encoder part.
-    decoding_func: list
+    decoding_func: tuple
         Activation functions for the decoder part.
     func_output: str
```
Comment (milesgranger) [inline gordo_components/model/factories/lstm_autoencoder.py]: I know you didn't change this, but just noticed it now; this parameter doesn't exist. Should be renamed to `out_func` I suppose.

File: gordo_components/model/factories/lstm_autoencoder.py [hunk 7]
```diff
@@ -117,8 +111,8 @@ def lstm_symmetric(
     n_features: int,
     n_features_out: int = None,
     lookback_window: int = 1,
-    dims: Tuple[int, ...] = (256, 128, 64),
-    funcs: Tuple[str, ...] = ("tanh", "tanh", "tanh"),
+    dims: Tuple[int, ...] = (),
+    funcs: Tuple[str, ...] = (),
```
Comment (milesgranger) [inline gordo_components/model/factories/lstm_autoencoder.py]: Should have the default values here. 
Right now, all of our `Keras..` models can be initialized and _"just work"_ by supplying the `kind` parameter (and we'll eventually have a default for that as well, to be more scikit-learn players) 

However, removing defaults here will cause the construction of the model to fail because of the first check in this very function.

File: gordo_components/model/factories/utils.py [hunk 8]
```diff
@@ -15,3 +16,13 @@ def hourglass_calc_dims(compression_factor, encoding_layers, n_features):
         round(n_features - (i * average_slope)) for i in range(1, encoding_layers + 1)
     ]
     return dims
+
+
+def check_dim_func_len(prefix: str, dim: Tuple[int, ...], func: Tuple[str, ...]):
```
Comment (milesgranger) [inline gordo_components/model/factories/utils.py:45]: docs plz

File: gordo_components/model/factories/utils.py [hunk 9]
```diff
@@ -15,3 +16,13 @@ def hourglass_calc_dims(compression_factor, encoding_layers, n_features):
...
+    if len(dim) != len(func):
+        raise ValueError(
+            f"The length (i.e. the number of network layers) of {prefix}_dim "
+            f"({len(dim)}) and {prefix}_func ({len(func)}) must be equal. If only "
+            f"{prefix}_dim or {prefix}_func was passed, ensure that its length matches "
+            "the that of the {prefix} parameter not passed."
```
Comment (milesgranger) [inline gordo_components/model/factories/utils.py]: "the that of the"

File: tests/gordo_components/model/test_lstm_autoencoder.py [hunk 10]
```diff
@@ -18,8 +19,13 @@ def test_lstm_symmetric_checks_dims(self):
         """
         Test that lstm_symmetric validates parameter requirements
         """
+        # Raise an error if empty dimension and function layers defined.
         with self.assertRaises(ValueError):
-            lstm_symmetric(4, dims=[], funcs=[])
+            lstm_symmetric(4, dims=(), funcs=())
+
```
Comment (milesgranger) [inline tests/gordo_components/model/test_lstm_autoencoder.py:44]: Additionally, you might also guarantee that something like `lstm_symmetric(4)` would work. As mentioned in a previous comment, a bit strange that the default args would illicit a `ValueError` in this function.

File: gordo_components/model/factories/lstm_autoencoder.py (hunk 6 above)
Comment (ryanjdillon) [reply gordo_components/model/factories/lstm_autoencoder.py]: It seems like `out_func` is used in the derivative functions where one only designates the function to be used and the dimensions are calculated from the inputs. 

Here it seems like `decoding_func` is being used [here](https://github.com/ryanjdillon/gordo-components/blob/dimension_function_error_checking/gordo_components/model/factories/lstm_autoencoder.py#L89-L92), just getting pulled out as `activation`.

```python
 89     # decoding layers
 90     for i, (n_neurons, activation) in enumerate(zip(decoding_dim, decoding_func)):
 91         return_seq = i != len(decoding_dim) - 1  # Don't return sequences on 2nd to last
 92         model.add(LSTM(n_neurons, activation=activation, return_sequences=return_seq))
```

File: gordo_components/model/factories/lstm_autoencoder.py (hunk 7 above)
Comment (ryanjdillon) [reply gordo_components/model/factories/lstm_autoencoder.py]: Thanks Miles, I missed this one. Strange the tests passed... I'll take a look there and see if I can add something that would make that fail next time.

File: gordo_components/model/factories/lstm_autoencoder.py [hunk 11]
```diff
@@ -41,16 +44,16 @@ def lstm_model(
         One timestep = current observation in the sample.
         Two timesteps = current observation + previous observation in the sample.
         ...
-    encoding_func: list
+    encoding_dim: tuple
+        Tuple of numbers with the number of neurons in the encoding part.
```
Comment (EinarJohnsen) [inline gordo_components/model/factories/lstm_autoencoder.py:48]: Minor detail, but here the documentation is using period at the end and in the ```gordo_components/model/factories/feedforward_autoencoder.py```, the documentation for encoding/decoding does not use a period at the end. Maybe we should change to one or the other?

Comment (ryanjdillon) [reply gordo_components/model/factories/lstm_autoencoder.py:48]: Grammatically speaking, since these aren't complete sentences, there shouldn't be a period; though, I am not sure what is considered standard practice here.

File: tests/gordo_components/model/test_lstm_autoencoder.py (hunk 10 above)
Comment (ryanjdillon) [reply tests/gordo_components/model/test_lstm_autoencoder.py:44]: I ended up landing on comparing the each layer of the different LSTM models' (with deafult parameters) configurations via the `layer.get_config()` method. 

This then ensure that the models pass default config muster and also that they are the same with default config.

File: gordo_components/model/factories/lstm_autoencoder.py (hunk 11 above)
Comment (milesgranger) [reply gordo_components/model/factories/lstm_autoencoder.py:48]: We've talked about this a bit, and I agree we haven't settled and it's a bit _lazy_ at times. However, we strive to follow the numpy style of documentation, which tends to be written in sentence format. But I think we're all guilty of slipping here and there.

Comment (ryanjdillon) [reply gordo_components/model/factories/lstm_autoencoder.py:48]: Sounds good to me :+1: I love the numpy docstring format, and if complete sentences are sanctioned by them then I think that is more then enough reason to do that.

Comment (ryanjdillon) [reply gordo_components/model/factories/lstm_autoencoder.py:48]: It doesn't seem most of these are written as complete sentences. I'll just give it a whack and see if you all agree with my rephrasing of things.

Comment (ryanjdillon) [reply gordo_components/model/factories/lstm_autoencoder.py:48]: In taking a look at things, it seems like numpy does the same as we are doing here.

[Theirs](https://github.com/numpy/numpy/blob/master/numpy/core/numeric.py#L123):
```
out : ndarray
    Array of zeros with the same shape and type as `a`.
```

Ours:
```
encoding_dim: tuple
    Tuple of numbers with the number of neurons in the encoding part.
```

So here they are using an incomplete sentence with a period. Erik likes the period as it indicated the end of the definition, which I agree is a useful thing. 

In this case, I'll leave the definitions all the same, and just add `.`s where missing in those files I've handled here in this PR.

File: gordo_components/model/factories/lstm_autoencoder.py [hunk 12]
```diff
@@ -71,17 +74,8 @@ class (e.x. Adam(lr=0.01,beta_1=0.9, beta_2=0.999)).  If no arguments are
...
-            f"must be equal for the encoder."
-        )
-
-    if len(decoding_dim) != len(decoding_func):
-        raise ValueError(
-            f"Number of layers ({len(decoding_dim)}) and number of functions ({len(decoding_func)}) "
-            f"must be equal for the decoder."
-        )
+    check_dim_func_len("encoding", encoding_dim, encoding_func)
+    check_dim_func_len("decoding", decoding_dim, decoding_func)
```
Comment (milesgranger) [inline gordo_components/model/factories/lstm_autoencoder.py:78]: Nice. :+1:

File: gordo_components/model/factories/lstm_autoencoder.py [hunk 13]
```diff
@@ -166,18 +160,19 @@ class (e.x. Adam(lr=0.01,beta_1=0.9, beta_2=0.999)).  If no arguments are
...
+        n_features_out=n_features_out,
+        lookback_window=lookback_window,
+        encoding_dim=dims,
+        decoding_dim=dims[::-1],
+        encoding_func=funcs,
+        decoding_func=funcs[::-1],
+        out_func=out_func,
+        optimizer=optimizer,
+        optimizer_kwargs=optimizer_kwargs,
+        compile_kwargs=compile_kwargs,
```
Comment (milesgranger) [inline gordo_components/model/factories/lstm_autoencoder.py:175]: I like the use of the keyword arguments. :1st_place_medal:

File: gordo_components/model/factories/utils.py [hunk 14]
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
+
+import math
+from typing import Tuple
+
+
+def hourglass_calc_dims(compression_factor, encoding_layers, n_features):
```
Comment (milesgranger) [inline gordo_components/model/factories/utils.py]: Need type hints, return type and numpy docs plz

File: gordo_components/model/factories/utils.py (hunk 8 above)
Comment (milesgranger) [reply gordo_components/model/factories/utils.py:45]: Numpy style docs.

File: tests/gordo_components/model/test_factories_utils.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        ((1, 3, 10), [10, 10, 10]),
+        ((0, 3, 100000), [66667, 33334, 1]),
+    ],
+)
+def test_hourglass_calc_dims_check_dims(test_input, test_expected):
+    """
+    Test that hourglass_calc_dims implements the correct dimensions
+    """
+    dims = hourglass_calc_dims(*test_input)
+    assert dims == test_expected
```
Comment (milesgranger) [inline tests/gordo_components/model/test_factories_utils.py:27]: LoooOoOoove it! :heart:

**Review by milesgranger — COMMENTED** [review]
Very nice sir, I think after you adjust the few comments I have on docs, you can rebase/squash this into 2-3 commits?

**Comment by ryanjdillon** [comment]
> Very nice sir, I think after you adjust the few comments I have on docs, you can rebase/squash this into 2-3 commits?

All doc'ed and squash'ed :+1:

File: gordo_components/model/factories/utils.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+def check_dim_func_len(prefix: str, dim: Tuple[int, ...], func: Tuple[str, ...]):
+    """
+    Check that the number of layer dimensions and layer functions are equal
+
+    Parameters
+    ----------
+    prefix: str
+        Parameter name prefix for error message generation (Options: "encoding" or "decoding").
+    dim: tuple of int
```
Comment (milesgranger) [inline gordo_components/model/factories/utils.py:53]: `Tuple[int, ...]`

**Comment by milesgranger** [comment]
You're also going to have to reword your first two commits so they aren't clipped. <~50 chars I think.

**Review by milesgranger — APPROVED** [review]
LGTM
